### PR TITLE
`CommitContext` records accessed input keys

### PR DESCRIPTION
**Pull request by fabiotudone-da**
Enables `CommitContext` to record and output (without iteration requirements) the input keys that were accessed.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 1]
```diff
@@ -37,7 +41,10 @@ private[kvutils] trait CommitContext {
   /** Retrieve value from output state, or if not found, from input state. */
   def get(key: DamlStateKey): Option[DamlStateValue] =
     outputs.get(key).orElse {
-      inputs.getOrElse(key, throw Err.MissingInputState(key))
+      val value = inputs.getOrElse(key, throw Err.MissingInputState(key))
+      if (accessedInputKeys.add(key))
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:45]: We don't need to keep track of the read order just the set of read keys. I.e., we could make the resulting `Iterable` deterministic by simply using a `SortedSet`.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Insertion order should be deterministic too (because access order is deterministic due to pre-execution and post-execution being deterministic on a given submission) and is slightly more efficient with a `LinkedHashSet`. @andreaslochbihler-da could you validate my reasoning?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala [hunk 2]
```diff
@@ -28,14 +28,7 @@ class CommitContextSpec extends WordSpec with Matchers {
...
 
-    "record accessed input keys" in {
-      val context = newInstance(Map(aKey -> Some(aValue)))
-      context.get(aKey)
-      context.getAccessedInputKeys.size shouldBe 1
-      context.getAccessedInputKeys.head shouldBe aKey
-    }
-
-    "maintain the order of accessed input keys based on when they were accessed first" in {
+    "records all accessed input keys and iterates over them in a stable order" in {
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala:31]: @miklos-da This really checks all the properties we're interested in and it includes the previous one, so I removed the latter.

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala:32]: That's fine. In order to document by a test that the input set is not necessarily the same as the read keys I would suggest adding a test case where you have 2 input keys but only read one (even if they're `None`, doesn't matter).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala (hunk 1 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Also, not sure if we really need deterministic or stable iteration over the input keys at all; do we already know that?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala (hunk 2 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala:32]: Good idea, added.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala (hunk 1 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: We don't even need deterministic order for the read set, we only care about membership and nothing else in the end.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: So no reliance by any ledger on any of these properties? If so yes and I could also just use an `HashSet` and save some more memory.

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Not that I know of. It's easy to fix this in the actual execution engine implementation if needed.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Done, can you have a further look?

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Looks good to me.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Thanks!

Comment (andreaslochbihler-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: For pre-execution, we must compute the same serialization for the read set on all replicas. This is because the read set contributes to the hash the replicas need to each consensus on. So we will ultimately need a deterministic order on the keys in the read set. For now, using any `Set` is fine as long as we recover determinism before/during serialization, e.g., by ordering all keys in it.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: OK, then we might want to revisit this because always iterating in access order is cheap here and spares us reordering later. Added a note to the LLD and will also cross-link this discussion.
